### PR TITLE
feat(promote): add "promotion" feature to gui

### PR DESCRIPTION
**Pull request by stohrendorf**

### CODE REVIEW

**Comment by zyv** [comment]
If repository state is older than deployments (forgot to fetch), then upon pressing `p` comes an exception:

```
zaytsev@work:~/src/django_insurance_api_v2$ ghd gui
Traceback (most recent call last):
  File "/home/zaytsev/src/ghd/docker/ghd.py", line 14, in <module>
    run_main()
  File "/home/zaytsev/src/ghd/docker/ghd.py", line 10, in run_main
    main_group()
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/zaytsev/src/ghd/docker/commands/utils.py", line 22, in wrapper
    return asyncio.run(f(*args, **kwargs))
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/zaytsev/src/ghd/docker/commands/__init__.py", line 219, in cmd_gui
    await gui_main(gh)
  File "/home/zaytsev/src/ghd/docker/commands/gui.py", line 371, in gui_main
    await main.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 166, in on_input
    return await focused.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 166, in on_input
    return await focused.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 161, in on_input
    if any(await self.on[key.code or str(key)](self, key)):
  File "/home/zaytsev/src/ghd/docker/saint/signal.py", line 67, in __call__
    return await self.emit(*args, **kwargs)
  File "/home/zaytsev/src/ghd/docker/saint/signal.py", line 63, in emit
    results.append(await f)
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/tasks.py", line 608, in _wait_for_one
    return f.result()  # May raise f.exception().
  File "/home/zaytsev/src/ghd/docker/commands/gui.py", line 312, in show_promote_confirmation
    git_log = "\n".join(get_git_log(recent_deployment["ref"], ref)) if recent_deployment else ""
TypeError: can only join an iterable
```

**Comment by stohrendorf** [comment]
@zyv The problem you described should be resolved now.  An appropriate message is shown if the commit diff could not be determined.

**Comment by mm-matthias** [comment]
I'm not available for an in-depth review at the moment.

**Comment by zyv** [comment]
Now I'm getting:

```
$ ghd gui
Traceback (most recent call last):
  File "/home/zaytsev/src/ghd/docker/ghd.py", line 14, in <module>
    run_main()
  File "/home/zaytsev/src/ghd/docker/ghd.py", line 10, in run_main
    main_group()
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zaytsev/.cache/pypoetry/virtualenvs/ghd--nsfylpF-py3.8/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/zaytsev/src/ghd/docker/commands/utils.py", line 22, in wrapper
    return asyncio.run(f(*args, **kwargs))
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/zaytsev/src/ghd/docker/commands/__init__.py", line 219, in cmd_gui
    await gui_main(gh)
  File "/home/zaytsev/src/ghd/docker/commands/gui.py", line 391, in gui_main
    await main.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 166, in on_input
    return await focused.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 166, in on_input
    return await focused.on_input(key)
  File "/home/zaytsev/src/ghd/docker/saint/widget.py", line 161, in on_input
    if any(await self.on[key.code or str(key)](self, key)):
  File "/home/zaytsev/src/ghd/docker/saint/signal.py", line 67, in __call__
    return await self.emit(*args, **kwargs)
  File "/home/zaytsev/src/ghd/docker/saint/signal.py", line 63, in emit
    results.append(await f)
  File "/home/zaytsev/opt/miniconda/envs/python-3.8/lib/python3.8/asyncio/tasks.py", line 608, in _wait_for_one
    return f.result()  # May raise f.exception().
  File "/home/zaytsev/src/ghd/docker/commands/gui.py", line 314, in show_promote_confirmation
    commits, completed = await self._gh.get_commits_until(ref, recent_deployment["ref"])
  File "/home/zaytsev/src/ghd/docker/github/__init__.py", line 128, in get_commits_until
    if commit["sha"] == until:
TypeError: string indices must be integers
```

File: docker/github/__init__.py [hunk 1]
```diff
@@ -115,6 +115,22 @@ async def get_deployment_statuses(self, deployment_id: int) -> list:
             reverse=True,
         )
 
+    async def get_commits(self, sha: str):
+        sha_arg = urlencode({"sha": sha})
+        return await self.get(f"/repos/{self.repo_path}/commits?{sha_arg}")
+
+    async def get_commits_until(self, sha: str, until: str):
+        sha_arg = urlencode({"sha": sha})
+        commits = await self.get(f"/repos/{self.repo_path}/commits?{sha_arg}")
```
Comment (zyv) [inline docker/github/__init__.py:124]: Macht das nicht Sinn statt code zu doppeln die Funktion oben hier zu verwenden:

```python
for commit in await self.get_commits(sha):
```

File: docker/commands/gui.py [hunk 2]
```diff
@@ -309,28 +309,34 @@ async def show_promote_confirmation(self, widget: Widget, key: blessed.keyboard.
...
+                author = committer["name"]
+                committed = localize_date(committer["date"])
+                message = commit["commit"]["message"].splitlines()[0]
+                sha = short_sha(commit["sha"])
+                return f"[{sha}  {committed}  {author}]  {message}"
+
+            if not commits:
+                git_log = colorama.Fore.CYAN + "No additional commits" + colorama.Fore.RESET
+            else:
+                git_log_lines = list(reversed(list(map(commit_to_oneline, commits))))
```
Comment (zyv) [inline docker/commands/gui.py]:
```suggestion
                git_log_lines = list(map(commit_to_oneline, commits))[::-1]
```
?

File: docker/commands/gui.py [hunk 3]
```diff
@@ -309,28 +309,34 @@ async def show_promote_confirmation(self, widget: Widget, key: blessed.keyboard.
...
-            )
+
+        if recent_deployment:
+            commits, completed = await self._gh.get_commits_until(ref, recent_deployment["ref"])
+
+            def commit_to_oneline(commit):
+                committer = commit["commit"]["committer"]
+                author = committer["name"]
+                committed = localize_date(committer["date"])
+                message = commit["commit"]["message"].splitlines()[0]
```
Comment (zyv) [inline docker/commands/gui.py]:
```suggestion
                message, *_ = commit["commit"]["message"].splitlines()
```

File: docker/commands/gui.py (hunk 2 above)
Comment (stohrendorf) [reply docker/commands/gui.py]: I'd like to see the list of changes from oldes to newest, so I can read the message from top to bottom without changing reading direction for the commit list just to read the list of commits ordered by date.

Comment (zyv) [reply docker/commands/gui.py]: The code is equivalent isn't it, it's just that IMO doesn't make sense to call `reversed` to produce a generator, which will be then instantly evaluated in a `list`, but rather reverse the list coming out of `map` with `[::-1]` directly producing a new list.

Comment (stohrendorf) [reply docker/commands/gui.py]: Agreed, somehow I missed that `[::-1]`.

**Pull request merged**